Add shapes

Elv13 opened the pull request:
This pull request add more `gears.shape`s and add dynamically generated examples

This slow down the compilation by 2-3 seconds

Elv13 commented:
So, this PR, beside more shapes, add an early way to add tested code examples to the documentation. Those code snippets are ran at compile time and must run clean (or kill the build). For now, only test that can be ran without Awesome are used. I am not sure having compile time Xephyr is a good idea. One possibility is to add a dummy drawable/drawin on an SVG surface for code that doesn't require "live" CAPI usage. But this will be later.

This is an early version. I first need to make it pass Travis (and generate the `.svg`)

psychon commented on lib/gears/shape.lua:
```diff
@@ -87,21 +97,41 @@ function module.infobubble(cr, width, height, corner_radius, arrow_size, arrow_p
+        cr:line_to(0            , height   )
+        cr:line_to(width        , height   )
+        cr:line_to(width        , 0        )
+    end
```
Do I understand it correctly that a negative `arrow_length` makes the arrow point to the other direction? If so, should that be mentioned in the docs?

Elv13 replied:
I think the last commit in this PR address this quite well

psychon commented on lib/gears/shape.lua:
```diff
@@ -195,3 +484,4 @@ function module.transform(shape)
 end
 
 return module
+-- kate: space-indent on; indent-width 4; replace-tabs on;
```
Huh? This doesn't have a vim modeline (weird...) and we now have kate mode lines?

I guess everyone should somehow use the `.editconfig` file that we have in awesome, I just don't know how. :-(
(Adding special comments for every editor isn't going to fly.)

psychon commented on lib/gears/shape.lua:
```diff
@@ -168,6 +220,243 @@ function module.isosceles_triangle(cr, width, height)
+-- @SHAPE_cross_EXAMPLE@
+--
+-- @param cr A cairo context
+-- @tparam number width The shape with
```
`with` -> `width` (also in other places in this file)

psychon commented on lib/gears/shape.lua:
```diff
@@ -168,6 +220,243 @@ function module.isosceles_triangle(cr, width, height)
+-- @tparam number corner_radius 
+function module.octogon(cr, width, height, corner_radius)
+    local corner_radius = corner_radius or math.min(10, math.min(width, height)/4)
+    local offset = math.sqrt( (corner_radius*corner_radius) / 2 )
```
Isn't this just `offset = corner_radius / math.sqrt(2)` (assuming `corner_radius` is positive)?
That doesn't really help me understand this (why not just `corner_radius` directly?), but at least feels easier to read.

Oh and: Why `math.min(10)` in the line before this? To me it looks like just `math.min(width,height/4)` should do.

psychon commented on lib/gears/shape.lua:
```diff
@@ -168,6 +220,243 @@ function module.isosceles_triangle(cr, width, height)
+-- @tparam number width The shape with
+-- @tparam number height The shape height
+function module.circle(cr, width, height)
+    local size = width < height and (width/2) or (height/2)
```
Please use `math.min(width, height) / 2`, I think that's easier to read.

psychon commented on lib/gears/shape.lua:
```diff
@@ -168,6 +220,243 @@ function module.isosceles_triangle(cr, width, height)
+function module.circle(cr, width, height)
+    local size = width < height and (width/2) or (height/2)
+    local w_offset, h_offset = (width - size*2)/2, (height - size*2)/2
+    cr:arc(w_offset + size, h_offset + size, size, 0, 2*math.pi)
```
Uhm... so the x-coordinate of the center of the circle is `w_offset + size = (width - size*2)/2 + size = width/2` and the y-coordinate is just `height/2` as well. Wouldn't it be easer to use `cr:arc(width / 2, height / 2, size, 0, 2*math.pi)`?

psychon commented on lib/gears/shape.lua:
```diff
@@ -168,6 +220,243 @@ function module.isosceles_triangle(cr, width, height)
+end
+
+--- A diagonal parallelogram with the bottom right corner at x=0 and top left
+-- at x=width.
```
Isn't (0, 0) the top-left corner? I think this should say that the top-left corner is at x=0 and the bottom right one is at x=width. Also, you could mention that the top right corner is at (width-thickness, 0), that might simplify understanding this. (Is it usual to call this a thickness? I have no idea...)

psychon commented on lib/gears/shape.lua:
```diff
@@ -168,6 +220,243 @@ function module.isosceles_triangle(cr, width, height)
+-- @tparam boolean br If the bottom right corner is rounded
+-- @tparam boolean bl If the bottom left corner is rounded
+-- @tparam number rad The corner radius
+function module.partially_rounded_rect(cr, width, height, tl, tr, br, bl, rad)
```
Should `rounded_rect` be replaced with a call to this, just so that the code exists only once? (Dunno if that makes much of a difference)
Also, please move the two closer together.

psychon replied:
I think this comment wasn't addressed yet, was it?

Elv13 replied:
I don't think it is worth merging, its simple code, it's a pointless micro optimization, but the rounded_rect one is called often when resizing clients with awful.shape, lets avoid a few pointless "if". It's 4 lines

psychon replied:
Ok.

psychon commented on tests/test-shape.lua:
```diff
@@ -0,0 +1,54 @@
+    else
+        cr:stroke_preserve()
+        cr:set_source_rgba(0.380392156863,0.505882352941,1,0.5)
+        cr:fill()
```
Doesn't this fill partly draw above the stroke, which should be visible thanks to transparency? I'd do it the other way around, first fill, then stroke (which doesn't have any transparency) to avoid this. Or is that on purpose?

psychon commented on tests/test-shape.lua:
```diff
@@ -0,0 +1,54 @@
+        cr:fill()
+    end
+
+    cr:translate(96)
```
Please use `cr:translate(96, 0)`.

psychon commented:
Nice idea. I don't mind the new shapes (I wonder how many shapes you still can come up with...) and having (1) tested usage examples and (2) images of these shapes in the docs is both nice.
I don't have any suggestions on how to integrate this more nicely in the build process (having to run tests just so that the installed documentation is complete is weird...), but it's a really nice idea. Thanks!

Elv13 commented:
```sh
CMake Error at CMakeLists.txt:271 (file):

  file COPY cannot set permissions on

  "/home/travis/build/awesomeWM/awesome/.build-testing-gce-c4a4845b-a5c0-4ca8-a50d-9bd047905b5c-x86_64-linux-gnu-4.8/doc/images/shape_arrow.svg"
```

Any idea on that error? Seem like the folder is created with a different user. I looked at `travis.yml`, but beside a couple of `sudo`, it doesn't seem to do anything special.

Btw, should we remove those `sudo` and use a local prefix and add it to `$PATH`, `$LUA_PATH` and `$XDG_*`? This would avoid this kind of issues.

Elv13 commented:
Umm, eventually I will have to figure out how to get these into the coverage _and_ running without the test framework...

@psychon So, is this good enough to merge now? I think the unaddressed comment are easy to understand once you see the images.

psychon commented on lib/gears/shape.lua:
```diff
@@ -125,11 +204,14 @@ function module.arrow(cr, width, height, head_width, shaft_width, shaft_length)
+-- @tparam number width The shape width
 -- @tparam number height The shape height
-function module.hexagon(cr, width, height)
+function module.hexagon(cr, width, height) --TODO need to new args to control top width
```
This (part of the first commit) adds a TODO...?

psychon commented on lib/gears/shape.back:
```diff
@@ -0,0 +1,449 @@
+---------------------------------------------------------------------------
```
Uhm, I think something went wrong with this commit ("add radial_progress shape"). It adds way more stuff than what the commit messages makes me expect.

Should this PR be split into two parts, one addings lots of new shapes (basically the first few commits) and another one handles the new "magic doc-images"?

(GitHub's interface changed, I can't comment at the bottom of this commit... Let's see where this comment ends up)

Elv13 replied:
Oops, git rebase fail... (note to myself, never use -A in aliases again)

psychon commented on tests/shape/test-shape.lua:
```diff
@@ -0,0 +1,36 @@
+
+-- This is normal when running test, it will run twice
+if (not filepath) or (not svgpath) or key then
+    os.exit(0)
```
At least print a warning that it doesn't do anything? Or, even better, don't run it twice?

Elv13 replied:
I will have to figure out how to write tests that run on both the test run and the pre-processing run. One give coverage and the other the documentation. This will be improved in future commits.

psychon replied:
Then I would suggest: Don't run this in the tests at all. If something breaks, the build should still fail. The only missing piece would be code coverage. If you want, feel free to open an issue about that and assign it to me (it was me who created this code-coverage stuff, after all).

Elv13 replied:
Ok, I pushed one last version for today, I am exhausted and I will make even more silly mistakes if I keep rebasing this stuff. The test (should) no longer run, I removed this code.

psychon commented on docs/generate_examples.cmake:
```diff
@@ -0,0 +1,48 @@
+execute_process(COMMAND lua -e print\(package.path\) OUTPUT_VARIABLE LUA_PATH_)
+set(ENV{LUA_PATH} "${SOURCE_DIR}/lib/?;${SOURCE_DIR}/lib/?.lua;${LUA_PATH_}")
+
+execute_process(COMMAND mkdir -p "${BUILD_DIR}/doc/images")
```
Can `file(MAKE_DIRECTORY "${BUILD_DIR}/doc/images")` handle this instead (e.g. so that no new process is forked)?

psychon commented on docs/generate_examples.cmake:
```diff
@@ -0,0 +1,48 @@
+    if (NOT SHAPE_ERROR STREQUAL "")
+        message(${SHAPE_ERROR})
+        message(${SHAPE_NAME} " SVG generation failed, bye")
+        exit(1)
```
Does this work? I can't find `exit` in the CMake docs. Shouldn't this be `message(FATAL_ERROR the rest of the message here)`?
(And for the previous line: `message(SEND_ERROR ${SHAPE_ERROR})`)

Elv13 replied:
Umm, good catch, . It was working, yes, but not the way it was intended to (Aka, it did call `message(FATAL_ERROR)` "indirectly".

psychon commented:
Also, could you push this as a branch to the awesomeWM/awesome repo? That way, awesome-robot should provide a link to the finished docs to (hopefully) look at.

psychon commented on docs/generate_examples.cmake:
```diff
@@ -0,0 +1,47 @@
+    execute_process(
+        COMMAND lua ${SOURCE_DIR}/tests/shape/test-shape.lua ${SHAPE_FILE} ${SHAPE_SVG}
+        OUTPUT_VARIABLE SHAPE_OUTPUT
+        ERROR_VARIABLE  SHAPE_ERROR
```
`execute_process` has `RESULT_VARIABLE`, according to the docs. It sounds like this will (sometimes?!?) be set to the exit code of the process. That should catch Lua errors more reliable (`lua -e 'error()'` has a non-zero exit code).

Elv13 replied:
Currently. if there is anything on stderr, it exit. I think it is what we want, even if there would be a "0" exit code. No?

psychon replied:
Yup. I was thinking that there could be errors that are only indicated by the exit code, but... you are right, that doesn't make any sense.

psychon commented:
Hm. Apparently there is no easy way to look at the resulting web pages except for cloning the git and checking out the branch. Anyway, here is the result from that travis build:

https://github.com/awesomeWM/apidoc/commits/add_shapes-491f17f

Compare view: https://github.com/awesomeWM/apidoc/compare/45a8a04...e0f16a9

Comparison against master (gh-pages): https://github.com/awesomeWM/apidoc/compare/gh-pages...e0f16a9

Elv13 commented:
I git cloned the page, it display fine locally

psychon commented:
Yup, looks great here as well. Nice work!
The docs for the parallelogram still say "A diagonal parallelogram with the bottom right corner at x=0 and top left at x=width.", but it's the bottom left corner which is at x=0 and the top right one which is at x=width.
Also, "A stroke only rounded bar with only percent set. This path is not closed." is hard to understand. How about "A partial rounded bar. How much of the rounded bar is visible depends on the given percentage value. Note that this shape is not closed and thus filling it doesn't make much sense." (by the way, what's the point of this shape? Perhaps some usage example should also be mentioned in the docs for it.)
I think that all the usage examples begin with an empty line. I guess that should be removed (either by editing the source or via some more CMake magic...).

psychon commented on tests/shape/rectangular_tag.lua:
```diff
@@ -0,0 +1,13 @@
+shape.rectangular_tag(cr, 70, 70)
+show(cr) --DOC_HIDE
+ 
+-- shape.rectangular_tag(cr,20,70) --TODO broken
```
This one shows up in the generated API docs (and sounds like something that should be fixed).

Elv13 replied:
I will add a DOC_HIDE for now, it has a minor visual issue, nothing terrible. I will fix it later when I improve the svg generator to show more than 3 images

Elv13 commented:
Ok, I will do that tomorrow. Can I merge it after fixing that? I have another (actually, 5) PR that re-use this framework coming for other modules, so I this has to be merged before I submit it. The coverage issue can be addressed later (it would be nice to have it, if not every new commits I will make will regress the coverage % as I plan to always include tested code examples where relevant.

psychon commented:
The decreased code coverage isn't an issue for me. The code is actually covered, but the tool just misses that. That doesn't worry me much.
I'd still like to take another look at this tomorrow. Sorry.
(P.S.: I'll be AFK next week, so I no reviews from me then.)

psychon commented:
After looking at this again: I should be less of a control freak and you are doing fine work. Please take a careful look at the final commits tomorrow (mostly to spot other rebase errors). Afterwards, feel free to merge. Thanks!
